File: getstarted/helloworld-braket/tdpnet.js
const {Manager} = require('socket.io-client')

class Session{
	constructor(){
	}

	connect(host){
		let manager = new Manager(host)
		this.socket = manager.socket('/')
		this.socket.io.on('error', data => this.onError(data))
		this.socket.on('connect', () => this.onConnected(Date.now()))
		this.socket.on('granted', data => this.onGranted(data))
		this.socket.on('denied', data => this.onDenied(data))
		this.socket.on('data', data => this.onData(data, this.valid(data)))
		this.socket.on('signal', data => this.onSignal(data, this.valid(data)))
	}

	onError(_){}
	onConnected(_){}
	onGranted(_){}
	onDenied(_){}
	onData(data, valid){ this.onCommand(data, valid, false)}
	onSignal(data, valid){ this.onCommand(data, valid, true)}
	onCommand(data,valid, signal){}

	valid(data){
		return data.from != undefined
	}

	send(id, data){
		this.socket.emit(id, data)
	}

	signin(data){
		this.send('signin', data)
	}

	signoff(data){
		this.send('signoff', data)
	}

    message(from, to, subject, detail){
        return {
            timestamp: Date.now(),
            from,
            to,
            subject,
            detail
        }
    }
    
}

const http = require('http')
const xpr = require('express')
const socket = require('socket.io')
const {v4 : uuid} = require('uuid')

class Desk{
	constructor(config){
		this.server = xpr()
		this.server.use(xpr.static(config.home))

		this.listen(config.port)
	}

	listen(port){
		let app = http.createServer(this.server)

		this.dispatch(socket(app))

		app.listen(port, () => this.onListening(port))
	}

	dispatch(io){
		io.on('connection', socket => {
			let id = this.getSessionId()
			let session = this.createSession(socket, id)

			if (session)
				socket.on('disconnect', () => this.onCloseSession(session, id))
			else
				socket.disconnect()
		})
	}

	onListening(){}

	onCloseSession(_){}
	createSession(_){}
	getSessionId(){
		return uuid()
	}

}

module.exports = {
    Session,
    Desk
}

File: getstarted/python/helloworld-sp/tdpnet.py
import socketio
import uuid
from http.server import BaseHTTPRequestHandler, HTTPServer
from socketserver import ThreadingMixIn
import os
import time

class Session:
    def __init__(self):
        self.socket = None

    def connect(self, host):
        self.manager = socketio.Client()
        self.manager.on('error', self.on_error)
        self.manager.on('connect', lambda: self.on_connected())
        self.manager.on('granted', self.on_granted)
        self.manager.on('denied', self.on_denied)
        self.manager.on('data', lambda data: self.on_data(data, self.valid(data)))
        self.manager.on('signal', lambda data: self.on_signal(data, self.valid(data)))
        self.manager.connect(host)

    def on_error(self, data):
        pass

    def on_connected(self):
        pass

    def on_granted(self, data):
        pass

    def on_denied(self, data):
        pass

    def on_data(self, data, valid):
        self.on_command(data, valid, False)

    def on_signal(self, data, valid):
        self.on_command(data, valid, True)

    def on_command(self, data, valid, signal):
        pass

    def valid(self, data):
        return 'from' in data

    def send(self, id, data):
        self.manager.emit(id, data)

    def signin(self, data):
        self.send('signin', data)

    def signoff(self, data):
        self.send('signoff', data)

    def message(self, from_, to, subject, detail):
        return {
            'timestamp': int(time.time() * 1000),
            'from': from_,
            'to': to,
            'subject': subject,
            'detail': detail
        }


class Desk:
    def __init__(self, config):
        self.server = HTTPServer((config['host'], config['port']), self.RequestHandler)
        self.server.config = config
        self.server.listen = self.listen
        self.dispatch(socketio.Server())

    def listen(self, port):
        self.server.listen(port)

    class RequestHandler(BaseHTTPRequestHandler):
        def do_GET(self):
            root = os.path.join(self.server.config['home'])
            if self.path == '/':
                filename = root + '/index.html'
            else:
                filename = root + self.path
            self.send_response(200)
            self.end_headers()
            self.wfile.write(open(filename, 'rb').read())

    def dispatch(self, io):
        io.on('connection', self.handle_connection)

    def handle_connection(self, socket):
        id = self.get_session_id()
        session = self.create_session(socket, id)
        if session:
            socket.on('disconnect', lambda: self.on_close_session(session, id))
        else:
            socket.disconnect()

    def on_listening(self, port):
        pass

    def on_close_session(self, session, id):
        pass

    def create_session(self, socket, id):
        pass

    def get_session_id(self):
        return str(uuid.uuid4())

